modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java

```diff
@@ -0,0 +1,367 @@
+                    WatchEvent evt = (WatchEvent) cursorDesc.getKey().next();
+
+                    List<SingleEntryResponse> resp = new ArrayList<>(evt.entryEvents().size() * 2);
+
+                    for (EntryEvent e : evt.entryEvents()) {
+                        Entry o = e.oldEntry();
+
+                        Entry n = e.entry();
+
+                        resp.add(new SingleEntryResponse(o.key(), o.value(), o.revision(), o.updateCounter()));
+
+                        resp.add(new SingleEntryResponse(n.key(), n.value(), n.revision(), n.updateCounter()));
+                    }
+
+                    clo.result(new MultipleEntryResponse(resp));
+                }
+            }
+            else if (clo.command() instanceof CursorCloseCommand) {
+                CursorCloseCommand cursorCloseCmd = (CursorCloseCommand) clo.command();
+
+                IgniteBiTuple<Cursor<?>, CursorType> val = cursors.get(cursorCloseCmd.cursorId());
+
+                if (val == null) {
+                    clo.result(null);
+
+                    return;
+                }
+
+                try {
+                    val.getKey().close();
+                }
+                catch (Exception e) {
+                    throw new IgniteInternalException(e);
+                }
+
+                clo.result(null);
+            }
+            else if (clo.command() instanceof WatchRangeKeysCommand) {
+                WatchRangeKeysCommand watchCmd = (WatchRangeKeysCommand) clo.command();
+
+                IgniteUuid cursorId = new IgniteUuid(UUID.randomUUID(), 0L);
+
+                Cursor<WatchEvent> cursor =
+                    storage.watch(watchCmd.keyFrom(), watchCmd.keyTo(), watchCmd.revision());
+
+                cursors.put(cursorId, new IgniteBiTuple<>(cursor, CursorType.WATCH));
+
+                clo.result(cursorId);
+            }
+            else if (clo.command() instanceof WatchExactKeysCommand) {
+                WatchExactKeysCommand watchCmd = (WatchExactKeysCommand) clo.command();
+
+                IgniteUuid cursorId = new IgniteUuid(UUID.randomUUID(), 0L);
+
+                Cursor<WatchEvent> cursor = storage.watch(watchCmd.keys(), watchCmd.revision());
+
+                cursors.put(cursorId, new IgniteBiTuple<>(cursor, CursorType.WATCH));
+
+                clo.result(cursorId);
+            }
+            else
+                assert false : "Command was not found [cmd=" + clo.command() + ']';
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onSnapshotSave(String path, Consumer<Throwable> doneClo) {
+        // Not implemented yet.
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onSnapshotLoad(String path) {
+        // Not implemented yet.
+        return false;
+    }
+
+    /** */
+    private static Condition toCondition(ConditionInfo info) {
+        byte[] key = info.key();
+
+        ConditionType type = info.type();
+
+        if (type == ConditionType.KEY_EXISTS)
```

Comment:
`Switch(type)` construction will be more natural and more effective here

Reply:
Agreed, but this is actually not my code. 
Class was renamed by me. 
Let's fix it by another ticket.